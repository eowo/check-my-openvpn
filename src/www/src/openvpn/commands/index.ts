export * from "./log";
export * from "./pid";
export * from "./status";
export * from "./bytecount";
