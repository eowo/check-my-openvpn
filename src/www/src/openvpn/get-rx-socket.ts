import { Socket } from "net";
import { Observable, from, Observer, fromEvent, Subject, race } from "rxjs";
import { tap, scan, filter, map, mergeMap, mapTo } from "rxjs/operators";
import {
  compose,
  converge,
  head,
  isEmpty,
  concat,
  last,
  init,
  tail,
  unnest,
  split
} from "ramda";

const isCompletePackage = compose<ReadonlyArray<string>, any, boolean>(
  isEmpty,
  last
);

const collectPackage = (acc: string[], cur: string[]) => {
  if (isCompletePackage(acc)) acc = [];
  return !isEmpty(acc)
    ? unnest([init(acc), [concat(last(acc), head(cur))], tail(cur)])
    : concat(acc, cur);
};

const createObservableSocket: (socket: Socket) => Observable<string> = socket =>
  Observable.create((observer: Observer<string | Error>) => {
    fromEvent(socket, "data")
      .pipe(
        map((data: Uint8Array) => data.toString()),
        tap(console.log),
        map(split(/\n/)),
        scan(collectPackage, []),
        filter(isCompletePackage),
        map((packages: string[]) => init(packages)),
        mergeMap((packages: string[]) => from(packages))
      )
      .subscribe((msg: string) => observer.next(msg));
    fromEvent(socket, "error").subscribe((error: Error) =>
      observer.error(error)
    );
    fromEvent(socket, "close").subscribe((hadError: boolean) =>
      hadError
        ? observer.error("Socket was closed due to a transmission error")
        : observer.complete()
    );
  });

const createObserverSocket = (socket: Socket) => {
  const s = new Subject<string>();
  s.subscribe({ next: cmd => socket.write(cmd) });
  return s;
};

export const getRxSocket = (source: Observable<Socket>) =>
  new Observable<[Observable<string>, Subject<string>]>(observer =>
    source
      .pipe(
        filter(socket => socket !== undefined),
        mergeMap(socket =>
          race(
            fromEvent<string>(socket, "connect").pipe(mapTo(socket)),
            fromEvent<Error>(socket, "error").pipe(mapTo(undefined))
          )
        ),
        filter((socket: Socket | undefined) => socket !== undefined),
        map(
          converge((read, write) => [read, write], [
            createObservableSocket,
            createObserverSocket
          ])
        )
      )
      .subscribe({
        next([read, write]) {
          observer.next([read, write]);
        },
        error(err) {
          observer.error(err);
        },
        complete() {
          observer.complete();
        }
      })
  );
