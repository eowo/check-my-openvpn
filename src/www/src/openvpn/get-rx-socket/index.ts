export * from "./get-rx-socket";
export { ObservableSocketRead } from "./create-observable-socket-read";
export { ObservableSocketWrite } from "./create-observable-socket-write";
