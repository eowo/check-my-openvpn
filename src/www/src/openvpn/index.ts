export * from "./openvpn";
export { Commands } from "./get-commands";
