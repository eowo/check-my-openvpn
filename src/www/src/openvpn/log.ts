import { Observable } from "rxjs";
import { filter } from "rxjs/operators";
import { test } from "ramda";

export const log: ([read, send]) => Observable<string> = ([read, send]) =>
  Observable.create(observer => {
    send.next("log on\r\n");
    read.pipe(filter(test(/^>LOG:/))).subscribe(observer);
  });
