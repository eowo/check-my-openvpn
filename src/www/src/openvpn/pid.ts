import { Observable, Subject } from "rxjs";
import { filter, map } from "rxjs/operators";
import { test, compose, match, last } from "ramda";

export const pid: ([read, send]: [
  Observable<string>,
  Subject<string>
]) => Observable<string> = ([read, send]) =>
  new Observable(observer => {
    send.next("pid\r\n");
    read
      .pipe(
        filter(test(/^SUCCESS: pid=/)),
        map(
          compose<string, string[], string>(
            last,
            match(/pid=(\d*)/)
          )
        )
      )
      .subscribe(observer);
  });
