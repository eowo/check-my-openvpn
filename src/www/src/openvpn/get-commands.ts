import { Observable } from "rxjs";
import { map } from "rxjs/operators";
import { converge } from "ramda";
import { logEnable, status, pid } from "./commands";
import { log } from "./real-time-messages";
import { ObservableSocketWrite, ObservableSocketRead } from "./get-rx-socket";

export interface Commands {
  logEnable: (on: true) => Observable<string>;
  log: Observable<string>;
  pid: Observable<string>;
  status: Observable<string[]>;
}

export const getCommands = () => (
  source: Observable<[ObservableSocketRead, ObservableSocketWrite]>
): Observable<Commands> =>
  new Observable(observer =>
    source
      .pipe(
        map(
          converge(
            (logEnable, pid, status, log) => ({
              logEnable,
              pid,
              status,
              log
            }),
            [logEnable, pid, status, log]
          )
        )
      )
      .subscribe(observer)
  );
