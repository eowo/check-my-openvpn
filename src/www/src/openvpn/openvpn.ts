import { createConnection, Socket } from "net";
import {
  fromEvent,
  Observable,
  of,
  race,
  ReplaySubject,
  Subscriber,
  throwError
} from "rxjs";
import {
  catchError,
  mapTo,
  mergeMap,
  take,
  tap,
  timeout
} from "rxjs/operators";
import { Commands, getCommands } from "./get-commands";
import { getRxSocket } from "./get-rx-socket";

const connect = (storedSocket: ReplaySubject<Socket>) => (
  host: string,
  port: number
) =>
  new Observable((observer: Subscriber<Commands>) => {
    of(createConnection({ host, port }))
      .pipe(
        mergeMap((socket: Socket) =>
          race(
            fromEvent<string>(socket, "connect").pipe(
              take(1),
              mapTo(socket),
              timeout(3000),
              catchError((error) => {
                socket.destroy();
                return throwError(error);
              })
            ),
            fromEvent<Error>(socket, "error").pipe(
              mergeMap((error) => throwError(error))
            )
          )
        ),
        tap((socket: Socket) => storedSocket.next(socket)),
        getRxSocket(),
        getCommands()
      )
      .subscribe(observer);
  });

export interface OpenVPN {
  connect: (host: string, port: number) => Observable<Commands>;
  disconnect: () => void;
}

export const openVpn = (): OpenVPN => {
  const storedSocket = new ReplaySubject<Socket>(1);

  return {
    connect: connect(storedSocket),
    disconnect: () =>
      storedSocket.pipe(take(1)).subscribe({
        next: (socket) => socket.end()
      })
  };
};
