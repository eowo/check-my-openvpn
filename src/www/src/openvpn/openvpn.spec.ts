import { createConnection } from "net";
import { TimeoutError } from "rxjs";
import { openVpn } from "./openvpn";

jest.mock("net");

describe("openVpn", () => {
  let socket: any;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    socket = mockSocket();
    (createConnection as jest.Mock).mockReturnValue(socket);
  });

  it("should have connect and destroy properties", () => {
    expect(openVpn()).toMatchSnapshot();
  });

  describe("connect()", () => {
    it("should be able to connect and emit commands", (done) => {
      socket.addEventListener.mockImplementation(
        (event: string, listener: () => void) =>
          event === "connect" && listener()
      );

      openVpn()
        .connect("10.8.0.1", 5555)
        .subscribe({
          next: (commands) => {
            expect(createConnection).toHaveBeenCalledWith({
              host: "10.8.0.1",
              port: 5555
            });
            expect(commands).toMatchSnapshot();
          },
          complete: () => done()
        });
    });

    it("should handle error event correctly", (done) => {
      socket.addEventListener.mockImplementation(
        (event: string, listener: (error: Error) => void) =>
          event === "error" && listener(new Error("Error happend"))
      );

      openVpn()
        .connect("10.8.0.1", 5555)
        .subscribe({
          error: (error) => {
            expect(error).toEqual(Error("Error happend"));
            done();
          }
        });
    });

    it("should handle tiemout error", (done) => {
      jest.useFakeTimers();

      openVpn()
        .connect("10.8.0.1", 5555)
        .subscribe({
          error: (error) => {
            expect(socket.destroy).toHaveBeenCalled();
            expect(error).toEqual(new TimeoutError());
            done();
          }
        });

      jest.advanceTimersByTime(3000);
      jest.useRealTimers();
    });
  });

  describe("disconnect()", () => {
    it("should call socket.end()", (done) => {
      socket.addEventListener.mockImplementation(
        (event: string, listener: () => void) =>
          event === "connect" && listener()
      );
      socket.end.mockImplementation(() => done());

      const o = openVpn();
      o.connect("10.8.0.1", 5555).subscribe({ complete: () => o.disconnect() });
    });
  });
});

const mockSocket = () => ({
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
  destroy: jest.fn(),
  end: jest.fn()
});
