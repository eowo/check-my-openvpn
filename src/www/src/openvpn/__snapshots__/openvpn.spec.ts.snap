// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openVpn connect() should be able to connect and emit commands 1`] = `
Object {
  "bytecount": Observable {
    "_isScalar": false,
    "operator": MapOperator {
      "project": [Function],
      "thisArg": undefined,
    },
    "source": Observable {
      "_isScalar": false,
      "operator": MapOperator {
        "project": [Function],
        "thisArg": undefined,
      },
      "source": Observable {
        "_isScalar": false,
        "operator": MapOperator {
          "project": [Function],
          "thisArg": undefined,
        },
        "source": Observable {
          "_isScalar": false,
          "operator": FilterOperator {
            "predicate": [Function],
            "thisArg": undefined,
          },
          "source": Observable {
            "_isScalar": false,
            "source": Subject {
              "_isScalar": false,
              "closed": false,
              "hasError": false,
              "isStopped": false,
              "observers": Array [],
              "thrownError": null,
            },
          },
        },
      },
    },
  },
  "bytecountRequest": [Function],
  "loadStats": Observable {
    "_isScalar": false,
    "operator": SwitchMapOperator {
      "project": [Function],
    },
    "source": Observable {
      "_isScalar": false,
      "operator": TimeoutWithOperator {
        "absoluteTimeout": false,
        "scheduler": AsyncScheduler {
          "SchedulerAction": [Function],
          "actions": Array [],
          "active": false,
          "now": [Function],
          "scheduled": undefined,
        },
        "waitFor": 1000,
        "withObservable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
      },
      "source": Observable {
        "_isScalar": false,
        "operator": FilterOperator {
          "predicate": [Function],
          "thisArg": undefined,
        },
        "source": Observable {
          "_isScalar": false,
          "operator": CatchOperator {
            "caught": [Circular],
            "selector": [Function],
          },
          "source": Observable {
            "_isScalar": false,
            "operator": MapToOperator {
              "value": true,
            },
            "source": Observable {
              "_isScalar": false,
              "operator": MergeMapOperator {
                "concurrent": Infinity,
                "project": [Function],
              },
              "source": Observable {
                "_isScalar": false,
                "operator": FilterOperator {
                  "predicate": [Function],
                  "thisArg": undefined,
                },
                "source": Observable {
                  "_isScalar": false,
                  "_subscribe": [Function],
                },
              },
            },
          },
        },
      },
    },
  },
  "log": Observable {
    "_isScalar": false,
    "operator": CatchOperator {
      "caught": [Circular],
      "selector": [Function],
    },
    "source": Observable {
      "_isScalar": false,
      "operator": FilterOperator {
        "predicate": [Function],
        "thisArg": undefined,
      },
      "source": Observable {
        "_isScalar": false,
        "source": Subject {
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
      },
    },
  },
  "logEnable": [Function],
  "pid": Observable {
    "_isScalar": false,
    "operator": SwitchMapOperator {
      "project": [Function],
    },
    "source": Observable {
      "_isScalar": false,
      "operator": TimeoutWithOperator {
        "absoluteTimeout": false,
        "scheduler": AsyncScheduler {
          "SchedulerAction": [Function],
          "actions": Array [],
          "active": false,
          "now": [Function],
          "scheduled": undefined,
        },
        "waitFor": 1000,
        "withObservable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
      },
      "source": Observable {
        "_isScalar": false,
        "operator": FilterOperator {
          "predicate": [Function],
          "thisArg": undefined,
        },
        "source": Observable {
          "_isScalar": false,
          "operator": CatchOperator {
            "caught": [Circular],
            "selector": [Function],
          },
          "source": Observable {
            "_isScalar": false,
            "operator": MapToOperator {
              "value": true,
            },
            "source": Observable {
              "_isScalar": false,
              "operator": MergeMapOperator {
                "concurrent": Infinity,
                "project": [Function],
              },
              "source": Observable {
                "_isScalar": false,
                "operator": FilterOperator {
                  "predicate": [Function],
                  "thisArg": undefined,
                },
                "source": Observable {
                  "_isScalar": false,
                  "_subscribe": [Function],
                },
              },
            },
          },
        },
      },
    },
  },
  "status": Observable {
    "_isScalar": false,
    "operator": SwitchMapOperator {
      "project": [Function],
    },
    "source": Observable {
      "_isScalar": false,
      "operator": TimeoutWithOperator {
        "absoluteTimeout": false,
        "scheduler": AsyncScheduler {
          "SchedulerAction": [Function],
          "actions": Array [],
          "active": false,
          "now": [Function],
          "scheduled": undefined,
        },
        "waitFor": 1000,
        "withObservable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
      },
      "source": Observable {
        "_isScalar": false,
        "operator": FilterOperator {
          "predicate": [Function],
          "thisArg": undefined,
        },
        "source": Observable {
          "_isScalar": false,
          "operator": CatchOperator {
            "caught": [Circular],
            "selector": [Function],
          },
          "source": Observable {
            "_isScalar": false,
            "operator": MapToOperator {
              "value": true,
            },
            "source": Observable {
              "_isScalar": false,
              "operator": MergeMapOperator {
                "concurrent": Infinity,
                "project": [Function],
              },
              "source": Observable {
                "_isScalar": false,
                "operator": FilterOperator {
                  "predicate": [Function],
                  "thisArg": undefined,
                },
                "source": Observable {
                  "_isScalar": false,
                  "_subscribe": [Function],
                },
              },
            },
          },
        },
      },
    },
  },
  "version": Observable {
    "_isScalar": false,
    "operator": SwitchMapOperator {
      "project": [Function],
    },
    "source": Observable {
      "_isScalar": false,
      "operator": TimeoutWithOperator {
        "absoluteTimeout": false,
        "scheduler": AsyncScheduler {
          "SchedulerAction": [Function],
          "actions": Array [],
          "active": false,
          "now": [Function],
          "scheduled": undefined,
        },
        "waitFor": 1000,
        "withObservable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
      },
      "source": Observable {
        "_isScalar": false,
        "operator": FilterOperator {
          "predicate": [Function],
          "thisArg": undefined,
        },
        "source": Observable {
          "_isScalar": false,
          "operator": CatchOperator {
            "caught": [Circular],
            "selector": [Function],
          },
          "source": Observable {
            "_isScalar": false,
            "operator": MapToOperator {
              "value": true,
            },
            "source": Observable {
              "_isScalar": false,
              "operator": MergeMapOperator {
                "concurrent": Infinity,
                "project": [Function],
              },
              "source": Observable {
                "_isScalar": false,
                "operator": FilterOperator {
                  "predicate": [Function],
                  "thisArg": undefined,
                },
                "source": Observable {
                  "_isScalar": false,
                  "_subscribe": [Function],
                },
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`openVpn should have connect and destroy properties 1`] = `
Object {
  "connect": Any<Function>,
  "disconnect": Any<Function>,
  "events": Any<AnonymousSubject>,
}
`;
